qutip/solver/multitraj.py

```diff
@@ -0,0 +1,464 @@
+from .. import Qobj, QobjEvo
+from .options import SolverOptions, SolverOdeOptions
+from .result import Result, MultiTrajResult
+from .parallel import get_map
+from time import time
+from .solver_base import Solver
+import numpy as np
+from copy import copy
+
+
+class MultiTrajSolver:
+    """
+    Basic class for multi-trajectories evolutions.
+
+    As :class:`Solver` it can ``run`` or ``step`` evolution.
+    It manage the random seed for each trajectory.
+
+    The actual evolution is done by a single trajectory solver::
+        ``_traj_solver_class``
+
+    Parameters
+    ----------
+    rhs : Qobj, QobjEvo
+        Right hand side of the evolution::
+            d state / dt = rhs @ state
+
+    options : SolverOptions
+        Options for the solver
+    """
+    _traj_solver_class = None
+    name = "generic multi trajectory"
+    optionsclass = SolverOptions
+    odeoptionsclass = SolverOdeOptions
+
+    def __init__(self, rhs, c_ops=(), *, options=None):
+        self.rhs = rhs
+        self._c_ops = c_ops
+        self.stats = {}
+        self.options = options
+        self.seed_sequence = np.random.SeedSequence()
+        self.traj_solver = False
+        self.result = None
+
+    def _read_seed(self, seed, ntraj):
+        """
+        Read user provided seed(s) and produce one for each trajectories.
+        Let numpy raise error for input that cannot be seeds.
+        """
+        if seed is None:
+            seeds = self.seed_sequence.spawn(ntraj)
+        elif isinstance(seed, np.random.SeedSequence):
+            seeds = copy(seed).spawn(ntraj)
+        elif not isinstance(seed, list):
+            seeds = np.random.SeedSequence(seed).spawn(ntraj)
+        elif isinstance(seed, list) and len(seed) >= ntraj:
+            seeds = [
+                seed_  if isinstance(seed_, np.random.SeedSequence)
+                else np.random.SeedSequence(seed_)
+                for seed_ in seed[:ntraj]
+            ]
+        else:
+            raise ValueError("A seed list must be longer than ntraj")
+        return seeds
+
+    def start(self, state0, t0, ntraj=1, seed=None, *, safe_ODE=None):
```

Comment:
`mesolve` uses `rho0`, `sesolve` and `mcsolve` use `psi0`.
But kets are accepted where most density matrix are, and operator input are accepted.
I have not passed to unify the documentation, but it will be needed.
I would go for simply `state` everywhere in this PR, with the plan to take another look at all solver's documentation later.